### PR TITLE
Reset discovery nodes in all transport node actions request.

### PR DESCRIPTION
**Pull request by SwethaGuptha**
### Description
Continuation to https://github.com/opensearch-project/OpenSearch/pull/14749, we don't required discovery nodes information in most of the transport actions extending TransportNodesActions, hence un-setting the discovery nodes object for such action before sending transport request to individual nodes this will prevent unnecessary serialization/deserialization of the discovery nodes object.

### Related Issues
Resolves #[Issue number to be closed when this PR is merged]
https://github.com/opensearch-project/OpenSearch/issues/14713

### Check List
- [ ] Functionality includes testing.
- [ ] API changes companion pull request [created](https://github.com/opensearch-project/opensearch-api-specification/blob/main/DEVELOPER_GUIDE.md), if applicable.
- [ ] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose), if applicable.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by SwethaGuptha** [comment]
> ❌ Gradle check result for [4114852](https://github.com/opensearch-project/OpenSearch/commit/411485285b58039290c25067d4baaba856ef1e60): [FAILURE](https://build.ci.opensearch.org/job/gradle-check/45018/)
> 
> Please examine the workflow log, locate, and copy-paste the failure(s) below, then iterate to green. Is the failure [a flaky test](https://github.com/opensearch-project/OpenSearch/blob/main/DEVELOPER_GUIDE.md#flaky-tests) unrelated to your change?

Flaky test https://github.com/opensearch-project/OpenSearch/issues/14294

File: server/src/main/java/org/opensearch/action/admin/cluster/node/reload/TransportNodesReloadSecureSettingsAction.java [hunk 1]
```diff
@@ -216,11 +216,8 @@ private boolean isNodeTransportTLSEnabled() {
     }
 
     private boolean isNodeLocal(NodesReloadSecureSettingsRequest request) {
-        if (null == request.concreteNodes()) {
-            resolveRequest(request, clusterService.state());
-            assert request.concreteNodes() != null;
-        }
-        final DiscoveryNode[] nodes = request.concreteNodes();
+        final DiscoveryNode[] nodes = resolveRequest(request, clusterService.state());
```
Comment (rajiv-kv) [inline server/src/main/java/org/opensearch/action/admin/cluster/node/reload/TransportNodesReloadSecureSettingsAction.java:219]: If the `clusterService.state()` changes b/w the co-ordinator making the request and datanode processing the request - wouldn't this impact ?

File: server/src/test/java/org/opensearch/action/support/nodes/TransportClusterStatsActionTests.java [hunk 2]
```diff
@@ -33,48 +33,12 @@
 
 public class TransportClusterStatsActionTests extends TransportNodesActionTests {
 
-    /**
-     * By default, we send discovery nodes list to each request that is sent across from the coordinator node. This
```
Comment (rajiv-kv) [inline server/src/test/java/org/opensearch/action/support/nodes/TransportClusterStatsActionTests.java:37]: Can we add the test to assert for actions that require the DiscoveryNode to be set in request ?

File: server/src/main/java/org/opensearch/action/support/nodes/TransportNodesAction.java [hunk 3]
```diff
@@ -234,23 +239,16 @@ class AsyncAction {
...
+            this.concreteNodes = resolveRequest(request, clusterService.state());
+            this.responses = new AtomicReferenceArray<>(this.concreteNodes.length);
         }
 
         void start() {
             final DiscoveryNode[] nodes = this.concreteNodes;
+            // As we transfer the ownership of discovery nodes to route the request to into the AsyncAction class,
+            // we remove the list of DiscoveryNodes from the request. This reduces the payload of the request and improves
+            // the number of concrete nodes in the memory.
+            request.setConcreteNodes(null);
```
Comment (rajiv-kv) [inline server/src/main/java/org/opensearch/action/support/nodes/TransportNodesAction.java:251]: - Why is this required - given that `resolveRequest` is no longer setting the `concreteNodes`. 
- Can the concreteNodes be removed from request always ? There might be some actions which might require setting it in request for stable nodeId's

File: server/src/main/java/org/opensearch/action/support/nodes/TransportNodesAction.java [hunk 4]
```diff
@@ -202,11 +202,16 @@ protected NodeResponse nodeOperation(NodeRequest request, Task task) {
 
     /**
      * resolve node ids to concrete nodes of the incoming request
-     **/
-    protected void resolveRequest(NodesRequest request, ClusterState clusterState) {
+     *
+     * @return
+     */
+    protected DiscoveryNode[] resolveRequest(NodesRequest request, ClusterState clusterState) {
```
Comment (rajiv-kv) [inline server/src/main/java/org/opensearch/action/support/nodes/TransportNodesAction.java:208]: Is it Ok to introduce this breaking change ? What if plugins extend the class and are consuming this method.

File: server/src/main/java/org/opensearch/action/admin/cluster/node/reload/TransportNodesReloadSecureSettingsAction.java (hunk 1 above)
Comment (SwethaGuptha) [reply server/src/main/java/org/opensearch/action/admin/cluster/node/reload/TransportNodesReloadSecureSettingsAction.java:219]: In this specific case it shouldn't impact. From the clusterService.state() we are only identifying if the node on which the setting has to be applied is the local node. If the node is part of the cluster then the clusterService.state() will always contain the datanode on which the request needs to be executed and if the node is not part of the cluster then we don't want to process this transport request anyway.,

Comment (rajiv-kv) [reply server/src/main/java/org/opensearch/action/admin/cluster/node/reload/TransportNodesReloadSecureSettingsAction.java:219]: If the value of node_id is set as `_local` wouldn't it matter if the the resolution happens on co-ordinator vs datanode ? Likewise for the value `_master` (or) `_cluster_manager`. Could we add testcases to validate the same ?

File: server/src/main/java/org/opensearch/action/support/nodes/TransportNodesAction.java (hunk 3 above)
Comment (SwethaGuptha) [reply server/src/main/java/org/opensearch/action/support/nodes/TransportNodesAction.java:251]: > Why is this required - given that resolveRequest is no longer setting the concreteNodes.

We need this because there are few callers of TransportNodesAction which pass a specific list of nodes to which transport request has to be sent.
E.g. - In AsyncShardFetch we identify the nodes to fetch from based on the cache. 
https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/gateway/AsyncShardFetch.java#L179

>Can the concreteNodes be removed from request always ? There might be some actions which might require setting it in request for stable nodeId's

Yes. Currently except for TransportNodesReloadSecureSettingsAction no other transport node actions is using this. As part of this PR, we are also removing dependency of TransportNodesReloadSecureSettingsAction from  concreteNodes object.

Comment (rajiv-kv) [reply server/src/main/java/org/opensearch/action/support/nodes/TransportNodesAction.java:251]: Got it, so it is set either in the constructor (or) as part of resolveRequest.

I think we can put back the boolean flag `IncludeDiscoveryNodes` and reset to null conditionally. This is to avoid breaking any external plugins that might extend the BaseNodeRequest.

File: CHANGELOG.md [hunk 5]
```diff
@@ -5,6 +5,7 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 
 ## [Unreleased 2.x]
 ### Added
+- Unset DiscoveryNodes in all transport node actions request ([#15131](https://github.com/opensearch-project/OpenSearch/pull/15131))
```
Comment (shwetathareja) [inline CHANGELOG.md:8]: Move it in the end in the Added list

File: server/src/main/java/org/opensearch/action/admin/cluster/node/hotthreads/NodesHotThreadsRequest.java [hunk 6]
```diff
@@ -71,6 +71,7 @@ public NodesHotThreadsRequest(StreamInput in) throws IOException {
      */
     public NodesHotThreadsRequest(String... nodesIds) {
         super(nodesIds);
+        this.setIncludeDiscoveryNodes(false);
```
Comment (shwetathareja) [inline server/src/main/java/org/opensearch/action/admin/cluster/node/hotthreads/NodesHotThreadsRequest.java:74]: you are setting this explicitly as plugin can extend TransportNodesAction, trying to understand what if we would have set it false in TransportNodesAction itself

File: server/src/main/java/org/opensearch/action/support/nodes/TransportNodesAction.java [hunk 7]
```diff
@@ -202,13 +202,24 @@ protected NodeResponse nodeOperation(NodeRequest request, Task task) {
 
     /**
      * resolve node ids to concrete nodes of the incoming request
-     **/
+     *
+     * @return
+     */
```
Comment (shwetathareja) [inline server/src/main/java/org/opensearch/action/support/nodes/TransportNodesAction.java:207]: not needed

File: server/src/main/java/org/opensearch/action/support/nodes/TransportNodesAction.java [hunk 8]
```diff
@@ -202,13 +202,24 @@ protected NodeResponse nodeOperation(NodeRequest request, Task task) {
...
     protected void resolveRequest(NodesRequest request, ClusterState clusterState) {
         assert request.concreteNodes() == null : "request concreteNodes shouldn't be set";
         String[] nodesIds = clusterState.nodes().resolveNodes(request.nodesIds());
         request.setConcreteNodes(Arrays.stream(nodesIds).map(clusterState.nodes()::get).toArray(DiscoveryNode[]::new));
     }
 
+    protected DiscoveryNode[] resolveDiscoveryNodes(NodesRequest request, ClusterState clusterState) {
```
Comment (shwetathareja) [inline server/src/main/java/org/opensearch/action/support/nodes/TransportNodesAction.java:214]: call resolveRequest from this method instead of duplicating the code

File: server/src/test/java/org/opensearch/action/support/nodes/TransportClusterStatsActionTests.java (hunk 2 above)
Comment (shwetathareja) [reply server/src/test/java/org/opensearch/action/support/nodes/TransportClusterStatsActionTests.java:37]: this should be in TransportNodesAction

File: server/src/main/java/org/opensearch/action/admin/cluster/node/stats/NodesStatsRequest.java [hunk 9]
```diff
@@ -75,6 +76,7 @@ public NodesStatsRequest(StreamInput in) throws IOException {
      */
     public NodesStatsRequest(String... nodesIds) {
         super(nodesIds);
```
Comment (rajiv-kv) [inline server/src/main/java/org/opensearch/action/admin/cluster/node/stats/NodesStatsRequest.java:78]: could we just add a constructor to the BaseNodesRequest (concreteNodes, includeDiscoveryNodes)

File: server/src/main/java/org/opensearch/action/admin/cluster/node/hotthreads/NodesHotThreadsRequest.java (hunk 6 above)
Comment (rajiv-kv) [reply server/src/main/java/org/opensearch/action/admin/cluster/node/hotthreads/NodesHotThreadsRequest.java:74]: Without a flag, I think we will end up not setting the `nodes` for transport requests made by plugin also. Are we thinking on having a flag in the TransportNodesAction and reset by derived actions ?

File: server/src/main/java/org/opensearch/action/support/nodes/TransportNodesAction.java [hunk 10]
```diff
@@ -234,23 +247,18 @@ class AsyncAction {
             this.task = task;
             this.request = request;
             this.listener = listener;
-            if (request.concreteNodes() == null) {
-                resolveRequest(request, clusterService.state());
```
Comment (rajiv-kv) [inline server/src/main/java/org/opensearch/action/support/nodes/TransportNodesAction.java:238]: i think we cannot remove this block as the derived classes can override this protected method. Ref[1](https://github.com/rajiv-kv/OpenSearch/blob/0040f4b766459610fae3a0342f26d8f78735778e/server/src/test/java/org/opensearch/action/support/nodes/TransportNodesActionTests.java#L347)

Lets continue calling the old method `resolveRequest` and reset after the call based on the include flag.

File: server/src/main/java/org/opensearch/action/support/nodes/TransportNodesAction.java [hunk 11]
```diff
@@ -234,23 +247,18 @@ class AsyncAction {
...
 
+        void start() {
+            final DiscoveryNode[] nodes = this.concreteNodes;
             if (request.getIncludeDiscoveryNodes() == false) {
-                // As we transfer the ownership of discovery nodes to route the request to into the AsyncAction class, we
-                // remove the list of DiscoveryNodes from the request. This reduces the payload of the request and improves
+                // As we transfer the ownership of discovery nodes to route the request to into the AsyncAction class,
+                // we remove the list of DiscoveryNodes from the request. This reduces the payload of the request and improves
                 // the number of concrete nodes in the memory.
                 request.setConcreteNodes(null);
```
Comment (rajiv-kv) [inline server/src/main/java/org/opensearch/action/support/nodes/TransportNodesAction.java:260]: can the reset be done in constructor ?

File: server/src/main/java/org/opensearch/indices/store/TransportNodesListShardStoreMetadata.java [hunk 12]
```diff
@@ -176,7 +176,7 @@ public Request(StreamInput in) throws IOException {
         }
 
         public Request(ShardId shardId, String customDataPath, DiscoveryNode[] nodes) {
-            super(nodes);
+            super(false, nodes);
```
Comment (shwetathareja) [inline server/src/main/java/org/opensearch/indices/store/TransportNodesListShardStoreMetadata.java:179]: this change to send discoveryNodes was introduced for replica shard store fetch, we may have to keep it in this flow?

**Review by rajiv-kv — APPROVED** [review]

**Comment by SwethaGuptha** [comment]
> ❌ Gradle check result for [84f48d4](https://github.com/opensearch-project/OpenSearch/commit/84f48d49205cdc070844e2fe32c42665093b15d6): [FAILURE](https://build.ci.opensearch.org/job/gradle-check/46282/)
> 
> Please examine the workflow log, locate, and copy-paste the failure(s) below, then iterate to green. Is the failure [a flaky test](https://github.com/opensearch-project/OpenSearch/blob/main/DEVELOPER_GUIDE.md#flaky-tests) unrelated to your change?

Flaky test https://github.com/opensearch-project/OpenSearch/issues/14294, 

> ❌ Gradle check result for [84f48d4](https://github.com/opensearch-project/OpenSearch/commit/84f48d49205cdc070844e2fe32c42665093b15d6): [FAILURE](https://build.ci.opensearch.org/job/gradle-check/46282/)
> 
> Please examine the workflow log, locate, and copy-paste the failure(s) below, then iterate to green. Is the failure [a flaky test](https://github.com/opensearch-project/OpenSearch/blob/main/DEVELOPER_GUIDE.md#flaky-tests) unrelated to your change?

Flaky test https://github.com/opensearch-project/OpenSearch/issues/14302 and https://github.com/opensearch-project/OpenSearch/issues/14294

**Comment by shwetathareja** [comment]
@SwethaGuptha there is conflict in ChangeLog.md

**Comment by SwethaGuptha** [comment]
> @SwethaGuptha there is conflict in ChangeLog.md

Resolved.

**Comment by SwethaGuptha** [comment]
> ❕ Gradle check result for [e7d56f1](https://github.com/opensearch-project/OpenSearch/commit/e7d56f152ea43ef0277165b6e89b27529f06a367): [UNSTABLE](https://build.ci.opensearch.org/job/gradle-check/46559/)
> 
>     * **TEST FAILURES:**
> 
> 
> ```
>       1 org.opensearch.backwards.MixedClusterClientYamlTestSuiteIT.test {p0=snapshot.status/10_basic/Get missing snapshot status succeeds when ignoreUnavailable is true}
> ```
> 
> Please review all [flaky tests](https://github.com/opensearch-project/OpenSearch/blob/main/DEVELOPER_GUIDE.md#flaky-tests) that succeeded after retry and create an issue if one does not already exist to track the flaky failure.

Flaky test https://github.com/opensearch-project/OpenSearch/issues/14294

**Comment by SwethaGuptha** [comment]
> ❕ Gradle check result for [e7d56f1](https://github.com/opensearch-project/OpenSearch/commit/e7d56f152ea43ef0277165b6e89b27529f06a367): [UNSTABLE](https://build.ci.opensearch.org/job/gradle-check/46559/)
> 
>     * **TEST FAILURES:**
> 
> 
> ```
>       1 org.opensearch.backwards.MixedClusterClientYamlTestSuiteIT.test {p0=snapshot.status/10_basic/Get missing snapshot status succeeds when ignoreUnavailable is true}
> ```
> 
> Please review all [flaky tests](https://github.com/opensearch-project/OpenSearch/blob/main/DEVELOPER_GUIDE.md#flaky-tests) that succeeded after retry and create an issue if one does not already exist to track the flaky failure.

https://github.com/opensearch-project/OpenSearch/issues/14294

**Review by shwetathareja — APPROVED** [review]

**Review by shwetathareja — APPROVED** [review]
